### PR TITLE
feat: scaffold API v1 conventions and OpenAPI contract

### LINKED ISSUE #129: [P1] API v1 conventions: error envelope, pagination, filtering, idempotency
## Parent
Part of #128

## Problem
Before implementing endpoints, we need one canonical API design contract so agents and SDKs can rely on stable behavior.

## Scope
Define and document API-wide conventions for:
- RFC 9457-compatible problem details errors (`application/problem+json`)
- cursor pagination (`nextCursor`, `limit`)
- filter/sort query conventions
- idempotency semantics for write operations (`Idempotency-Key` behavior)
- stable timestamp and ID formatting

## Deliverables
- `docs/api/conventions.md` (single source of truth)
- reusable TypeScript types for success/error envelopes in `lib/api/`
- request/response examples for each pattern

## Acceptance Criteria
- [ ] Every API route in #128 can reference this spec without ambiguity
- [ ] Error schema includes `type`, `title`, `status`, `detail`, `instance`, and extension members
- [ ] Pagination conventions documented with examples and edge-case behavior
- [ ] Idempotency behavior documented for `POST`/`PATCH` including replay window and conflict handling

## Boundaries
- No route implementation in this issue
- No OpenAPI document authoring in this issue (covered separately)

### LINKED ISSUE #130: [P1] Author OpenAPI 3.1 spec scaffold for API v1 and add spec linting
## Parent
Part of #128

## Problem
We need a machine-readable contract for agent tooling, SDK generation, and regression-safe API evolution.

## Scope
- Create initial OpenAPI 3.1 document for API v1
- Include global components (auth scheme, problem details schema, pagination schema)
- Add unique `operationId` values for all planned v1 endpoints
- Add lint/validation step in CI

## Suggested Files
- `docs/api/openapi.v1.yaml`
- `.github/workflows/test.yml` (or dedicated API-spec workflow)
- optional lint config (`.redocly.yaml` or Spectral config)

## Acceptance Criteria
- [ ] OpenAPI document validates in CI
- [ ] All v1 endpoints from #128 are represented (even if some are marked `x-internal: planned` initially)
- [ ] Error responses use shared Problem Details schema component
- [ ] Security scheme for bearer API key is documented once and reused

## Boundaries
- No SDK generation in this issue
- No endpoint runtime implementation in this issue

### PR DESCRIPTION
**Pull request by phrazzld**
## Summary
Builds the Phase 0 API contract foundation for agent-first API v1: canonical conventions + OpenAPI 3.1 scaffold + CI lint gate.

References #128.
Closes #129.
Closes #130.

## Changes
- Added API-wide conventions doc with RFC 9457 problem details, cursor pagination, filter/sort rules, timestamp/ID formatting, and idempotency semantics.
- Added OpenAPI 3.1 scaffold covering all v1 planned endpoints from #128 with shared bearer auth, problem schema, pagination schema, and unique operationIds.
- Added reusable API contract helpers in `lib/api/contracts.ts` for problem details, pagination envelope, and idempotency constants.
- Added focused unit tests for the API contract helpers.
- Added `lint:openapi` script and wired OpenAPI lint into CI `test.yml`.
- Linked API docs from README.

### #129 API conventions
- [x] Every API route in #128 can reference this spec without ambiguity.
- [x] Error schema includes `type`, `title`, `status`, `detail`, `instance`, and extension members.
- [x] Pagination conventions documented with examples and edge-case behavior.
- [x] Idempotency behavior documented for POST/PATCH including replay window and conflict handling.

### #130 OpenAPI scaffold + lint
- [x] OpenAPI document validates in CI (`bun run lint:openapi` + workflow step).
- [x] All v1 endpoints from #128 are represented (marked `x-internal: planned`).
- [x] Error responses use shared Problem Details schema component.
- [x] Security scheme for bearer API key is documented once and reused.

## Manual QA
1. `bun run lint:openapi`
   - Expected: spec validates (warnings allowed by current default ruleset).
2. `bun test lib/api/__tests__/contracts.test.ts`
   - Expected: 9 passing tests.
3. `bun lint`
   - Expected: ESLint passes with 0 warnings.
4. `bun type-check`
   - Expected: TypeScript check passes.
5. `git grep -n "operationId" docs/api/openapi.v1.yaml`
   - Expected: unique operation IDs for all declared paths.

## What Changed
```mermaid
graph TD
  A[#128 Agent-first API v1 Epic] --> B[#129 Conventions Contract]
  A --> C[#130 OpenAPI Scaffold]
  B --> D[docs/api/conventions.md]
  C --> E[docs/api/openapi.v1.yaml]
  C --> F[CI OpenAPI Lint Gate]
  B --> G[lib/api/contracts.ts]
  G --> H[lib/api/__tests__/contracts.test.ts]
```

## Before / After
- Before: no canonical API conventions doc, no machine-readable OpenAPI contract for v1 endpoints, no CI guardrail for API contract drift.
- After: conventions and OpenAPI v1 scaffold exist as source-of-truth artifacts, all planned v1 endpoints are modeled, and CI now validates the API contract.

## Test Coverage
- Added: `lib/api/__tests__/contracts.test.ts`
  - `buildProblemDetails`
  - `buildPaginatedResponse`
  - idempotency constants contract
- Verified project gates:
  - `bun lint`
  - `bun type-check`
  - pre-push Vitest suite (all passing)
- Gap intentionally left (out of scope): runtime API endpoint behavior tests (to be added with endpoint implementation issues).

## Summary by CodeRabbit

* **Documentation**
  * Added API conventions for Heartbeat API v1 and a full OpenAPI v3.1 contract; README updated to reference API contract docs.
  * Added a retrospective note for Issue #128 describing scope and CI contract pattern.

* **Chores**
  * Added a CI OpenAPI lint step and a local lint command (lint:openapi).

* **Tests**
  * Added tests validating API contract helpers (problem details, pagination, idempotency constants).

## Before / After (Polish Pass)
- Before: pagination constraints were documented but not fully enforced at helper and OpenAPI page-array boundaries; two minor bot review threads remained unresolved.
- After: pagination invariants are enforced in both code and schema (buildPaginatedResponse limit guard + maxItems: 200 on paginated response arrays), regression tests cover invalid/boundary limits, and all review threads are resolved.

### CODE REVIEW

**Comment by phrazzld** [comment]
Context: this PR intentionally lands contract-first artifacts (conventions + OpenAPI scaffold) before runtime endpoint implementation. OpenAPI lint currently passes with warnings from Redocly default recommended rules (license/tag descriptions/explicit 4xx responses). Those are tracked as follow-up polish and do not block schema validity or CI gating.

File: docs/api/openapi.v1.yaml [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+        scopes:
+          type: array
+          items:
+            $ref: "#/components/schemas/ApiKeyScope"
+        projectSlugs:
+          anyOf:
+            - type: "null"
+            - type: array
+              items:
+                type: string
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply docs/api/openapi.v1.yaml]: Classification: style
Severity: low
Decision: fix now. Applied OpenAPI 3.1 nullable-array shorthand for consistency.
Change: Updated `ApiKeyMetadata.projectSlugs` to `type: [array, "null"]` with string items in `docs/api/openapi.v1.yaml` (commit e1d6592).
Verification: `bun run lint:openapi`, `bun lint`, `bun type-check`, `bun test lib/api/__tests__/contracts.test.ts`.

File: lib/api/contracts.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  page: CursorPagination;
+};
+
+export function buildProblemDetails(
+  input: ProblemDetailsInput,
+): ProblemDetails {
+  const { extensions, ...base } = input;
+  return {
+    ...base,
+    ...(extensions ?? {}),
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply lib/api/contracts.ts]: Classification: bug
Severity: high
Decision: fix now. Canonical Problem Details fields should not be overridable by extensions.
Change: Reversed spread order in `buildProblemDetails` so extensions are merged first and base fields win (`lib/api/contracts.ts`, commit e1d6592). Added regression test asserting extension keys cannot override canonical fields (`lib/api/__tests__/contracts.test.ts`).
Verification: `bun test lib/api/__tests__/contracts.test.ts`, pre-push Vitest suite (996 passing), `bun lint`, `bun type-check`.

File: docs/api/openapi.v1.yaml [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+          type: array
+          minItems: 1
+          items:
+            $ref: "#/components/schemas/ApiKeyScope"
+        projectSlugs:
+          anyOf:
+            - type: "null"
+            - type: array
+              items:
+                type: string
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply docs/api/openapi.v1.yaml]: Classification: style
Severity: low
Decision: fix now. Applied the same nullable-array shorthand in the request schema.
Change: Updated `CreateApiKeyRequest.projectSlugs` to `type: [array, "null"]` with string items in `docs/api/openapi.v1.yaml` (commit e1d6592).
Verification: `bun run lint:openapi`, `bun lint`, `bun type-check`, `bun test lib/api/__tests__/contracts.test.ts`.

File: lib/api/contracts.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+} & Record<string, unknown>;
+
+export type ProblemDetailsInput = {
+  type: string;
+  title: string;
+  status: number;
+  detail?: string;
+  instance?: string;
+  extensions?: Record<string, unknown>;
+};
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply lib/api/contracts.ts:19]: Classification: risk
Severity: medium
Decision: reject. Instead of requiring `detail`/`instance` in TypeScript, we aligned the published contract to RFC 9457 behavior where `detail` may be omitted for security-sensitive responses.
Change: Updated conventions wording and removed `detail` from OpenAPI `ProblemDetails.required` (`docs/api/conventions.md`, `docs/api/openapi.v1.yaml`, commit e1d6592). TS remains optional and is now contract-aligned.
Verification: `bun run lint:openapi`, `bun lint`, `bun type-check`, pre-push Vitest suite (996 passing).

File: docs/api/conventions.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+- `nextCursor = null` means terminal page
+- Invalid cursor returns `400` Problem Details
+- Cursor is opaque and must not be parsed by clients
+
+## Filtering and Sorting
+
+Common query conventions:
+
+- `projectSlug=<slug>`
+- `status=<up|degraded|down|investigating|resolved>`
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply docs/api/conventions.md]: Classification: style
Severity: medium
Decision: fix now. Clarified route-specific status enums to avoid ambiguity.
Change: Updated `docs/api/conventions.md` to split status filters into monitor-route statuses (`up|degraded|down`) and incident-route statuses (`investigating|identified|resolved`) in commit e1d6592.
Verification: `bun lint`, `bun type-check`.

File: lib/api/contracts.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (moomooskycow) [reply lib/api/contracts.ts:46]: Classification: bug
Severity: high
Decision: fix now.
Change: Reversed merge order in `buildProblemDetails` so extension members cannot override canonical Problem Details fields (`lib/api/contracts.ts`, commit e1d6592). Added regression test in `lib/api/__tests__/contracts.test.ts`.
Verification: `bun test lib/api/__tests__/contracts.test.ts`, pre-push Vitest suite (996 passing), `bun lint`, `bun type-check`.

**Comment by phrazzld** [comment]
## Review Follow-up (Bot Issue Comments)

Processed actionable findings from bot **issue comments** and corresponding review threads.

- Fixed: OpenAPI nullable-array consistency for `projectSlugs` (`ApiKeyMetadata` + `CreateApiKeyRequest`).
- Fixed: Prevented extension payload from overriding canonical Problem Details fields in `buildProblemDetails`.
- Fixed: Clarified status filter vocabulary by route class (monitor vs incident).
- Fixed: Added coverage for missing contract helper edge cases (no extensions, empty data + terminal cursor).
- Fixed: Added runtime validation for ProblemDetails `status` range (`100..599`) with regression test.
- Contract decision: kept `detail` optional in TypeScript and aligned docs/OpenAPI to RFC 9457-compatible omission for security-sensitive cases.

Verification:
- `bun run lint:openapi`
- `bun lint`
- `bun type-check`
- pre-push Vitest suite: **997 passing tests**

**Comment by phrazzld** [comment]
## PR Unblocked

**Conflicts**: none
**CI**: green (pre-push Vitest full suite passed: 997 tests)
**Reviews**: addressed and replied on all open review threads; all unresolved threads resolved via GraphQL

Changes shipped in this unblock pass:
- fixed ProblemDetails extension precedence to protect canonical fields
- aligned docs/OpenAPI/TS contract semantics for optional `detail`
- clarified status filter enums by resource class
- added missing edge-case tests (no extensions, terminal pagination, invalid status)

Ready for re-review.

File: docs/api/conventions.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+}
+```
+
+Required members:
+
+- `type`
+- `title`
+- `status`
+- `detail` (optional, omit intentionally for security-sensitive errors)
+- `instance`
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply docs/api/conventions.md:36]: Classification: style
Severity: low
Decision: fix now.
Change: Restructured the section into explicit Required vs Optional members in `docs/api/conventions.md` (`instance` required; `detail` optional for security-sensitive omission) in commit 1104b5f.
Verification: `bun lint`, `bun type-check`, pre-push Vitest suite (997 passing).

File: lib/api/contracts.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
+export const IDEMPOTENCY_KEY_HEADER = "Idempotency-Key";
+export const API_IDEMPOTENCY_REPLAY_WINDOW_HOURS = 24;
+
+export type ProblemDetails = {
+  type: string;
+  title: string;
+  status: number;
+  detail?: string;
+  instance?: string;
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply lib/api/contracts.ts]: Classification: risk
Severity: medium
Decision: fix now.
Change: Made `instance` required in both `ProblemDetails` and `ProblemDetailsInput` in `lib/api/contracts.ts` to match OpenAPI + conventions contract (commit 1104b5f).
Verification: `bun test lib/api/__tests__/contracts.test.ts`, `bun lint`, `bun type-check`, pre-push Vitest suite (997 passing).

**Comment by phrazzld** [comment]
Follow-up pass complete on latest commit `1104b5f`: addressed Greptile deltas (`instance` now required in TS contract types; conventions section split into Required vs Optional members), replied on both new threads, and resolved them.

File: docs/api/openapi.v1.yaml [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+    MonitorsPageResponse:
+      type: object
+      required: [data, page]
+      properties:
+        data:
+          type: array
+          items:
+            $ref: "#/components/schemas/Monitor"
+        page:
+          $ref: "#/components/schemas/PageInfo"
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply docs/api/openapi.v1.yaml:444]: Classification: contract-hardening
Severity: low
Decision: fix now. Added explicit page array bounds to match the documented pagination cap.
Change: Added maxItems: 200 to MonitorsPageResponse.data, ChecksPageResponse.data, and IncidentsPageResponse.data in docs/api/openapi.v1.yaml.
Verification: bun run lint:openapi, bun lint, bun type-check, bun test lib/api/__tests__/contracts.test.ts.

File: lib/api/contracts.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  limit: number,
+): PaginatedResponse<T> {
+  return {
+    data,
+    page: {
+      limit,
+      nextCursor,
+    },
+  };
+}
```
[the remark this answers is not in the document]
Comment (moomooskycow) [reply lib/api/contracts.ts:67]: Classification: contract-hardening
Severity: low
Decision: fix now. Pagination helper now enforces contract invariants at module boundary.
Change: Added integer range validation (1-200) in buildPaginatedResponse (lib/api/contracts.ts) and added regression tests for invalid and boundary limits in lib/api/__tests__/contracts.test.ts.
Verification: bun test lib/api/__tests__/contracts.test.ts, bun lint, bun type-check, bun run lint:openapi.
